Do not allocate memory on reverts with errors of small static encoding size.

ekpyron opened the pull request:
I guess lack of micro-optimization like this is what drives people to aggressively resort to inline assembly like in this case seen in https://github.com/ethereum/solidity/issues/13149

r0qs commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -3363,16 +3375,25 @@ void IRGeneratorForStatements::revertWithError(
 {
+	bool needsAllocation = true;
+	if (auto size = staticEncodingSize(_parameterTypes))
+		needsAllocation = *size > CompilerUtils::generalPurposeMemoryStart;
```
Maybe we could also add a test case to check the `needsAllocation` logic?

ekpyron replied:
We have semantics tests for raising all sorts of errors of different sizes - all of them still pass.

nikola-matic commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -3363,16 +3375,25 @@ void IRGeneratorForStatements::revertWithError(
 )
 {
+	bool needsAllocation = true;
+	if (auto size = staticEncodingSize(_parameterTypes))
```
```suggestion
	if (optional<size_t> size = staticEncodingSize(_parameterTypes))
```
The type is not mentioned anywhere in the statement (at least on the right hand side), so it would be nice to make it explicit in this case.

ekpyron replied:
Is that really so :-)?

nikola-matic commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -3363,16 +3375,25 @@ void IRGeneratorForStatements::revertWithError(
+		<?needsAllocation>
 		let <pos> := <allocateUnbounded>()
+		<!needsAllocation>
+		let <pos> := 0
```
I'm surprised this change is reflected nowhere in the tests, in other words, I assume `allocateUnbounded` always returns 0 in our tests, which would mean that some coverage is missing.

Also, if this makes things cheaper, wouldn't this also be reflected in the gas cost test? Or I guess not, if `pos` was/is always 0?

ekpyron replied:
Why would you assume that? Then the change would just be nonsense :-D.
``allocateUnbounded`` never returns zero.
This is most definitely reflected in our tests - the behaviour just doesn't change (it can't, otherwise, it'd be a bug...) and we don't have tests involving eligible errors that consume sufficiently much gas, s.t. this change should show up in the gas statistics. We generally don't have verbatim tests on complex bytecode, since those would just break all the time, but this is covered in at least 50, probably more semantics tests.

ekpyron replied:
I added a case with artificially boosted gas consumption now, though, first before the change, then showing the savings after the change.

ekpyron committed:
Do not allocate memory on reverts with small errors.

Do not allocate memory on reverts with small errors.

Workaround for StringLiteralType

Addition require error test coverage

ekpyron commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -3363,16 +3375,26 @@ void IRGeneratorForStatements::revertWithError(
 {
+	bool needsAllocation = true;
+	if (optional<size_t> size = staticEncodingSize(_parameterTypes))
+		if (ranges::all_of(_parameterTypes, [](auto const* type) { return type && type->isValueType(); }))
```
Just added the additional restriction of all error arguments being value types to make sure we don't run into weird static array arguments that end up calculating hashes in scratch space during ABI encoding. Value types alone are safe.

ekpyron commented:
### `ir-optimize-evm+yul`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|            brink |  **`-0.1% ✅`** |                |                |
|           colony |           `0%` |                |                |
|        elementfi |           `0%` |                |                |
|              ens |           `0%` | **`-0.71% ✅`** |           `0%` |
|           gnosis |                |                |                |
|              gp2 |           `0%` |          `+0%` |          `+0%` |
|  perpetual-pools |           `0%` |          `-0%` | **`+0.01% ❌`** |
|         prb-math | **`-0.98% ✅`** | **`-0.97% ✅`** |           `0%` |
|          uniswap |           `0%` |          `-0%` |          `-0%` |
| yield_liquidator |           `0%` |           `0%` |           `0%` |
|         zeppelin | **`-0.02% ✅`** | **`-0.02% ✅`** |          `-0%` |

This actually saves 1% code size on prb-math...

ekpyron commented:
Looking and thinking through the possible cases here again, this might still be problematic for this case:
```
error E(bytes32 x);
contract C {
    bytes s;
    function f() public {
        revert E(bytes32(s));
    }
}
```

since the abi encoding from storage bytes to fixed bytes may involve temporary hasing in scartch space - given that, I'll postpone merging this until after the release pending a closer look.

cameel commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -3363,16 +3376,26 @@ void IRGeneratorForStatements::revertWithError(
+	bool needsAllocation = true;
+	if (optional<size_t> size = staticEncodingSize(_parameterTypes))
+		if (ranges::all_of(_parameterTypes, [](auto const* type) { return type && type->isValueType(); }))
+			needsAllocation = *size + 4 > CompilerUtils::generalPurposeMemoryStart;
```
So, just to make sure I understand why this is memory safe:
- It's ok to overwrite free mem pointer because we're reverting and not allocating anything before then.
- It's ok to overwrite the zero slot since the condition above excludes any empty dynamic arrays passed in as arguments.
- It's not ok to overwrite the area where variables moved to memory might be because one of them might be `<pos>` or `<end>`.

Is that correct?

ekpyron replied:
Yes, all that's true - but the restriction of types is in the current version not strong enough to exclude all uses of the scratch space during the encoding step, so that still needs fixing (restricting to value types for both the types of the arguments *and* the types of the error signature should do that - pending a closer look, though).

cameel commented on Changelog.md:
```diff
@@ -15,6 +15,7 @@ Compiler Features:
  * SMTChecker: Properties that are proved safe are now reported explicitly at the end of the analysis. By default, only the number of safe properties is shown. The CLI option ``--model-checker-show-proved-safe`` and the JSON option ``settings.modelChecker.showProvedSafe`` can be enabled to show the full list of safe properties.
  * SMTChecker: Group all messages about unsupported language features in a single warning. The CLI option ``--model-checker-show-unsupported`` and the JSON option ``settings.modelChecker.showUnsupported`` can be enabled to show the full list.
  * Yul EVM Code Transform: If available, use ``push0`` instead of ``codesize`` to produce an arbitrary value on stack in order to create equal stack heights between branches.
+ * Yul IR Code Generation: Cheaper code for reverting with errors of a static small encoding size.
```
This needs to be moved to 0.8.21.

cameel reviewed (commented):
Everything seems fine here.

Only needs to be rebased and have the changelog entry moved to the new version.

cameel commented:
Ah, wait, I also need to think through the cases where the scratch space would be overwritten. So maybe not ok yet (but at least I did not notice any problems directly in the implementation).

ekpyron commented:
Yeah, it's **not** fine as it stands - but I actually wanted to use this PR for a "teach how to review codegen for correctness" session, so I haven't fixed it yet - but not sure we'll have the time for it.
I'll draft it for now in any case.

cameel commented:
> Looking and thinking through the possible cases here again, this might still be problematic for this case:

Anything involving type conversions seems safe because the conversion is performed before the revert function (at least on 0.8.20):

```yul
/// @src 0:97:98  "s"
let _1_slot := 0x00
let expr_12_slot := _1_slot
/// @src 0:89:99  "bytes32(s)"
let expr_13 := convert_bytes_to_fixedbytes_from_t_bytes_storage_to_t_bytes32(expr_12_slot)
/// @src 0:87:100  "E(bytes32(s))"
{
    let _2 := allocate_unbounded()
    mstore(_2, 8203157970705126157628513801238240545114148153221041971593256941893860720640)
    let _3 := abi_encode_tuple_t_bytes32__to_t_bytes32__fromStack(add(_2, 4) , expr_13)
    revert(_2, sub(_3, _2))
}
```

So the only problems could come from the encoding. With parameters being restricted to value types I guess this excludes any possibility of it messing with the scratch space?

ekpyron commented:
It's possible that the current version is already save since we don't have *implicit* conversion from ``bytes`` to fixed bytes and the conversion will thereby happen before the revert snipped, yes.

cameel commented:
Ok, maybe better to be too careful. In any case, I tried a bunch of cases but did not came up with anything that could break it. For example:
- We have implicit conversion from string literals to `bytes32` but limited to 32-byte strings so we're covered.
- I also played with calldata or storage values but to access a value type it must be in an array/slice/struct/mapping and you must use indexing or member access, which happens before encoding.
- We don't have pointers to value types so that's out too (and you'd have to artificially make it point at scratch space to mess with this mechanism anyway).
- Selectors are hashed at compilation time so not an issue either.

Just in case I rechecked which types are marked as value types in `Types.h` and the only one that looks suspicious is `InaccessibleDynamic`, but I doubt we have to worry about that one.

Looking at `ABIFunctions::tupleEncoder()` and the encoding helpers it calls, but it seems that with value types we're skipping all the cases where something weird could happen. Though to be honest, this is a sprawling set of functions and I don't know all its nook and crannies so there could always be something obscure and counterintuitive that I missed.

ekpyron commented:
Yeah - that's why my plan was to screenshare on this PR and walk through what to look for and think through during such a review - maybe we'll still find time for it eventually, let's see :-). But yeah, especially with restricting to value types for the argument types as well should be safe.

chriseth commented:
We could merge a version in case there are no arguments at all.

nikola-matic commented on test/cmdlineTests/require_with_error_ir/output:
```diff
@@ -186,7 +186,9 @@ object "C_36" {
             function require_helper_t_error_7_CustomError_t_rational_1_by_1_t_stringliteral_332c39dcd398ea34a48b871898d589f55fc4c7bce00562fb670c972e7e1b0720(condition , expr_15) {
                 if iszero(condition)
                 {
+
```
Really? It newlines the Whiskers conditionals?

cameel commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -226,17 +251,28 @@ std::string YulUtilFunctions::revertWithError(
+	bool needsAllocation = true;
+
+	if (std::optional<size_t> size = staticEncodingSize(_parameterTypes))
+		if (ranges::all_of(_parameterTypes, [](auto const* type) { return type && type->isValueType(); }))
```
We could assert that `type` is not `InaccessibleDynamic`, since that's actually a value type.

cameel commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -32,11 +32,36 @@
+			type = TypeProvider::stringMemory();
+		if (type->isDynamicallyEncoded())
+			return std::nullopt;
+		encodedSize += type->calldataHeadSize();
```
@nikola-matic Actually, I was wrong on the call about using `memoryHeadSize()` here. We're storing ABI encoded error arguments so it's the calldata size that's relevant. This is correct as is.

cameel commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -32,11 +32,36 @@
+	size_t encodedSize = 0;
+	for (auto* type: _parameterTypes)
+	{
+		if (dynamic_cast<StringLiteralType const*>(type))
+			// Workaround for StringLiteralType, which will fail an assert on the
+			// type->calldataHeadSize() call; thus, assign string memory type (array)
+			// to type, which was and is explicitly provided to revertWithError where
+			// applicable anyway.
+			type = TypeProvider::stringMemory();
```
I think it would be simpler to remove this bit and just assert that the function expects a value type (and then swap the conditions in `revertWithError()` to avoid calling it with non-value types, the way @nikola-matic suggested).

That's because, when (if?) we enable this optimization for dynamic types and literals we'll need to have the same if in `revertWithError()` anyway. Having it here does not tell the caller that literals are handled like strings because that bit of info does not get outside of the function. The caller will need to repeat that check.

Anyway, that's just a small simplification, so no big deal if it gets merged as is, but the `if` is unnecessary IMO. Unless the value it provides is reminding us that a string literal is a possibility here :) But for that we could just have a shorter comment.

cameel replied:
Wait, I looked again at #15058 and I think that the fact that we're even getting a string literal here might be a bug after all.

I assumed that `requireOrAssertFunction()` used to use string literal type for it's argument, but I see [it was `TypeProvider::stringMemory()`](https://github.com/ethereum/solidity/pull/15058/files#diff-331d761d82437590a3abfd03a44ddd6b0abc88b54095707f69005276f5d1ac19L277-L281).

To keep it unchanged, the function should be passing `{TypeProvider::stringMemory()}` as parameter types into `revertWithError()`, but [it's instead passing `{_messageType}`](https://github.com/ethereum/solidity/pull/15058/files#diff-331d761d82437590a3abfd03a44ddd6b0abc88b54095707f69005276f5d1ac19R285), which is the type of the argument expression, and not of the parameter defined in the error signature.

That's a small difference and did not break anything in the original PR, because the calldata encoding of a string and a string literal is the same. It's only coming up now because we suddenly care about dynamic types and a literal is not marked as one.

nikola-matic replied:
Ah, good point (latter) Kamil; I wouldn't necessarily say it was a bug, because the string literal would then end up being encoded as a bytes array, just like `stringMemory()`, but this way is definitely better and more elegant as it allows us to remove the `StringLiteral` check/hack in the `staticEncodingSize()` function. This then makes the first comment (i.e. swapping the `if`s) unnecessary :)

cameel replied:
It is being encoded as bytes array either way - there's no separate encoding defined for string literals in the ABI. I'm actually surprised that ABI encoder can even handle conversion *to* that type. The signature is `Error(string)`, not `Error(<unnamable string literal type>)`.

I remember that you mentioned back then you needed it to be the literal and I accepted the explanation, but now I'm not really sure why. It does not seem correct :) It definitely should be the string type.

cameel commented on Changelog.md:
```diff
@@ -5,6 +5,7 @@ Language Features:
 
 
 Compiler Features:
+ * Yul IR Code Generation: Cheaper code for reverting with errors of a static small encoding size.
```
```suggestion
 * Yul IR Code Generation: Cheaper code for reverting with errors of a small static encoding size.
```

cameel commented on test/libsolidity/semanticTests/errors/small_error_optimization.sol:
Like I mentioned on the call, would be good to add a semantic test that would cover a range of examples of error parameter types (unless we already have that covered). That should fit in a single case and still be quite compact (there are not that many possibilities), and we'd have empirical verification that none of those types breaks things.

cameel reviewed (commented):
Would be good to get a fresh set of benchmarks from external tests. I'd expect the PR to have a lot more impact now that it's on top of #15058.

stackenbotten commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -226,17 +245,31 @@ std::string YulUtilFunctions::revertWithError(
+	bool needsAllocation = true;
+
+	if (std::optional<size_t> size = staticEncodingSize(_parameterTypes))
+		if (ranges::all_of(_parameterTypes, [](auto const* type) {
```
Coding style error

cameel reviewed (dismissed):
The code looks fine now, but review fixes to be squashed.

While at it, would be nice to update the test expectations in the first commit, because they seem out of date. I was hoping to see the effect on gas in the diff, but that commit seems to have been generated even before #14843 so it's useless. It should be updated or squashed.

Finally, what's the impact on external tests? I really expected there to be more of it with recent changes.

nikola-matic commented:
> Finally, what's the impact on external tests? I really expected there to be more of it with recent changes.

I wouldn't be surprised if it had very little impact, as I'd assume people would have been reverting small errors via assembly (which is why we even implemented this optimization).

cameel commented on test/libsolidity/semanticTests/errors/small_error_optimization.sol:
```diff
@@ -0,0 +1,22 @@
+	}
+}
+// ----
+// f() -> FAILURE, hex"92bbf6e8"
+// gas irOptimized: 221918
+// gas irOptimized code: 42800
+// gas legacy: 233746
+// gas legacy code: 37400
+// gas legacyOptimized: 224864
+// gas legacyOptimized code: 34200
```
Is this really the cost before optimization as the description says? i don't see it being updated in the other commit so if this was the case, CI would be failing.

nikola-matic replied:
No, it was my bad, I ended up fixing up the first commit instead of the second and then squashed. This was the before and after:
```
  Running via Yul: 
  Expected result:
  // f() -> FAILURE, hex"92bbf6e8"
  // gas irOptimized: 222533
  // gas irOptimized code: 43200
  // gas legacy: 233746
  // gas legacy code: 37400
  // gas legacyOptimized: 224864
  // gas legacyOptimized code: 34200

  Obtained result:
  // f() -> FAILURE, hex"92bbf6e8"
  // gas irOptimized: 221918 [-615 (+0%)]
  // gas irOptimized code: 42800 [-400 (+0%)]
  // gas legacy: 233746
  // gas legacy code: 37400
  // gas legacyOptimized: 224864
  // gas legacyOptimized code: 34200
```

cameel replied:
ok. That's a nice improvement.

But you should still at least update the commit description if you're not going to fix it in the commit. Or just squash it, because if it's wrong, it's worse than useless.

nikola-matic replied:
Agreed. I'll just squash everything into one comment, and leave this comment here in case someone is curious and stumbles upon it.

r0qs commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -226,17 +245,33 @@ std::string YulUtilFunctions::revertWithError(
+				return type && type->isValueType();
+			})
+		)
+			needsAllocation = *size + 4 > CompilerUtils::generalPurposeMemoryStart;
```
out of curiosity, why plus 4?

cameel replied:
Size of error selector.

r0qs replied:
Got it. So maybe it would be nice a comment above it :)
